shared-bindings/msgpack/__init__.c

```diff
@@ -0,0 +1,391 @@
+        for (size_t i=0; i<len; i++) {
+            mp_obj_dict_store(d, _unpack(s), _unpack(s));
+        }
+        return MP_OBJ_FROM_PTR(d);
+    }
+    switch (code) {
+        case 0xc0: return mp_const_none;
+        case 0xc2: return mp_const_false;
+        case 0xc3: return mp_const_true;
+        case 0xc4:
+        case 0xc5:
+        case 0xc6: {
+            // bin 8, 16, 32
+            size_t size = read_size(s, code-0xc4);
+            vstr_t vstr;
+            vstr_init_len(&vstr, size);
+            byte *p = (byte*)vstr.buf;
+            read(s, p, size);
+            return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
+        }
+        case 0xca: {
+            float f;
+            read(s, &f, 4);
+            return mp_obj_new_float(f);
+        }
+        case 0xd0:
+            return MP_OBJ_NEW_SMALL_INT((int8_t)read_bytes(s, 1));
+        case 0xd1:
+            return MP_OBJ_NEW_SMALL_INT((int16_t)read_bytes(s, 2));
+        case 0xd2:
+            return MP_OBJ_NEW_SMALL_INT((int32_t)read_bytes(s, 4));
+        case 0xd9:
+        case 0xda:
+        case 0xdb: {
+            // str 8, 16, 32
+            size_t size = read_size(s, code-0xd9);
+            vstr_t vstr;
+            vstr_init_len(&vstr, size);
+            byte *p = (byte*)vstr.buf;
+            read(s, p, size);
+            return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+        }
+        case 0xde:
+        case 0xdf: {
+            // map 16 & 32
+            size_t len = read_size(s, code - 0xde + 1);
+            mp_obj_dict_t *d = MP_OBJ_TO_PTR(mp_obj_new_dict(len));
+            for (size_t i=0; i<len; i++) {
+                mp_obj_dict_store(d, _unpack(s), _unpack(s));
+            }
+            return MP_OBJ_FROM_PTR(d);
+        }
+        case 0xdc:
+        case 0xdd: {
+            // array 16 & 32
+            size_t len = read_size(s, code - 0xdc + 1);
+            mp_obj_tuple_t *t = MP_OBJ_TO_PTR(mp_obj_new_tuple(len, NULL));
+            for (size_t i=0; i<len; i++) {
+                t->items[i] = _unpack(s);
+            }
+            return MP_OBJ_FROM_PTR(t);
+        }
+        case 0xc1:      // never used
+        case 0xc7:      // ext 8
+        case 0xc8:      // ext 16
+        case 0xc9:      // ext 32
+        case 0xcb:      // float 64
+        case 0xcc:      // uint 8
+        case 0xcd:      // uint 16
+        case 0xce:      // uint 32
+        case 0xcf:      // uint 64
+        case 0xd3:      // int 64
+        case 0xd4:      // fixenxt 1
+        case 0xd5:      // fixenxt 2
+        case 0xd6:      // fixenxt 4
+        case 0xd7:      // fixenxt 8
+        case 0xd8:      // fixenxt 16
+        default:
+            mp_raise_ValueError(translate("no unpacker found"));
+    }
+}
+
+// Read msgpack encoded object from stream
+STATIC mp_obj_t mod_msgpack_unpack(mp_obj_t stream_obj) {
+    msgpack_stream_t s = get_stream(stream_obj, MP_STREAM_OP_READ);
+    return _unpack(&s);
```

Comment:
Please move all of the actual parsing into `shared-module/msgpack/__init__.c`. `shared-bindings` should only include docs via python stub after `//|` and argument parsing code. All of the actual logic should go into shared-module. This leads to a clear Python API and C API. 

```suggestion
    return common_hal_msgpack_unpack(stream_obj);
```

Reply:
Thanks for all your comments. I made all changes. 

1. Not sure if `WriteableBuffer` is the correct type here: `def pack(obj: Any, buffer: WriteableBuffer) -> None:`
2. The test still fails, apparently due to missing translations.
3. I don't know how to implement logic tests.